imap_processing/lo/l1b/lo_l1b.py

```diff
@@ -537,36 +526,70 @@ def find_valid_asc(
     valid_mask : np.ndarray
         Boolean mask indicating valid ASCs.
     """
-    # Start with all ASCs as valid
-    valid_mask = np.ones(len(closest_start_acq_indices), dtype=bool)
+    # Apply each validation check independently on full arrays
+    valid_indices = _check_valid_indices(science_to_spin_indices)
+    valid_spin_count = _check_sufficient_spins(science_to_spin_indices, spin_data)
+
+    # Combine only these two masks:
+    valid_mask = valid_indices & valid_spin_count
+
+    total_invalid = (~valid_mask).sum()
+    if total_invalid > 0:
+        logger.info(f"Dropping {total_invalid} invalid ASCs total")
+
+    return valid_mask
+
 
-    # Filter out epochs where no valid spin data exists (all time_diff < 0)
-    no_prior_spin = np.all(time_diff < 0, axis=1)
-    if no_prior_spin.any():
-        logger.warning(f"Dropping {no_prior_spin.sum()} ASCs with no prior spin data")
-        valid_mask &= ~no_prior_spin
+def _check_valid_indices(science_to_spin_indices: np.ndarray) -> np.ndarray:
+    """
+    Check that all matched spin indices are valid (non-negative).
 
-    # Filter out epochs with invalid indices
-    invalid_indices = closest_start_acq_indices < 0
+    Parameters
+    ----------
+    science_to_spin_indices : np.ndarray
+        Indices of closest spin acquisitions.
+
+    Returns
+    -------
+    valid_mask : np.ndarray
+        Boolean mask where True indicates a valid index.
+    """
+    invalid_indices = science_to_spin_indices < 0
     if invalid_indices.any():
-        logger.warning(
-            f"Dropping {invalid_indices.sum()} ASCs with invalid spin indices"
-        )
-        valid_mask &= ~invalid_indices
-
-    # Filter out ASCs with insufficient spins
-    valid_cycles = spin_data["num_completed"] == 28
-    # Only check cycles for valid indices
-    valid_indices = closest_start_acq_indices[valid_mask]
-    insufficient_spins = ~valid_cycles.values[valid_indices]
-    if insufficient_spins.any():
-        logger.warning(
-            f"Dropping {insufficient_spins.sum()} ASCs with fewer than 28 spins"
-        )
-        # Update the mask for remaining valid entries
-        temp_mask = np.copy(valid_mask)
-        temp_mask[valid_mask] &= ~insufficient_spins
-        valid_mask = temp_mask
+        logger.warning(f"Found {invalid_indices.sum()} ASCs with invalid spin indices")
+    return ~invalid_indices
+
+
+def _check_sufficient_spins(
+    science_to_spin_indices: np.ndarray, spin_data: xr.Dataset
+) -> np.ndarray:
+    """
+    Check that matched spin cycles have sufficient spins (28 completed).
+
+    Parameters
+    ----------
+    science_to_spin_indices : np.ndarray
+        Indices of closest spin acquisitions.
```

Comment:
Why do we need this array here? I feel like this adds a bit of confusion in terms of returning the mask based on a different mapping than the input spin values. I would have assumed that this function returns the boolean array `spin_data["num_completed"] == 28`, then after that you can index into that boolean array if you want outside of this function. `good_spin_arr[science_to_spin_indices[valid_idx]]`